keps/sig-node/3960-pod-lifecycle-sleep-action/README.md

```diff
@@ -0,0 +1,376 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes the addition of a new sleep action for the PreStop lifecycle hook in Kubernetes, allowing containers to pause for a specified duration before termination. This enhancement aims to provide a more straightforward way to manage graceful shutdowns and improve the overall lifecycle management of containers.
+
+## Motivation
+
+Currently, Kubernetes supports two types of actions for PreStop hooks: exec and httpGet. Although these actions offer flexibility, they often require additional scripting or custom solutions to achieve a simple sleep functionality. A built-in sleep action would provide a more user-friendly and native solution for scenarios where a container needs to pause before shutting down, such as:
+
+- Ensuring that the container gracefully releases resources and connections.
+- Allowing a smooth transition in load balancers or service meshes.
+- Providing a buffer period for external monitoring and alerting systems.
+
+### Goals
+
+- Allow containers to perform cleanup or shutdown actions before being terminated, by sleeping for a specified duration in the preStop hook.
+- Improve the overall reliability and availability of Kubernetes applications by providing a way for containers to gracefully terminate.
+
+### Non-Goals
+
+- This KEP does not aim to replace other Kubernetes features that can be used to perform cleanup actions, such as init containers or sidecar containers.
+- This KEP does not aim to provide a way to pause or delay pod termination indefinitely.
+
+## Proposal
+
+We propose adding a new sleep action for the PreStop hook, which will pause the container for a specified duration before termination. The API changes will include the following:
+
+- Extending the LifecycleHandler object to support a new Sleep field.
+- Adding a SleepAction object that includes a Duration field to specify the sleep period in seconds.
+
+### User Stories (Optional)
+
+#### Story 1
+As a Kubernetes user, I want to configure my container to sleep for a specific duration before termination so that it can gracefully release resources and connections.
+
+#### Story 2
+As a Kubernetes user, I want to configure my container to sleep before termination to ensure a smooth transition in my service mesh or load balancer.
+
+#### Story 3
+As a Kubernetes user, I want to configure my container to sleep before termination to provide a buffer period for external monitoring and alerting systems.
+
+### Risks and Mitigations
+
+One potential risk of adding a sleep action to the preStop hook is that it may increase the time it takes for a pod to terminate, which can affect the overall availability of the application.
```

Comment:
hmm, this reads as if the sleep can override the termination period, and this is not how it is documented

> The Pod's termination grace period countdown begins before the PreStop hook is executed, so regardless of the outcome of the handler, the container will eventually terminate within the Pod's termination grace period. 

https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks

Reply:
Yes, it can **not** override the termination period, and also it can not be more than the termination grace period.  Let me rethink the wording here

Reply:
Actually, I think this is documented below.
> It's recommended that users set the duration to a value that is less than or equal to the terminationGracePeriodSeconds specified in the pod's spec.

But yes, this sentence is actually about `terminationGracePeriodSeconds`. Actually I can't really think of some risks associated with this feature. Do you have any thoughts?

Reply:
We need to make sure that whatever they set here, the `terminationGracePeriodSeconds` wins and terminates if it hits first.

We could statically validate that this is <= `terminationGracePeriodSeconds` but grace period can be reduced at runtime, so that's a partial solution.